ml-agents/mlagents/tf_utils/tf.py

```diff
@@ -28,3 +35,179 @@ def set_warnings_enabled(is_enabled: bool) -> None:
     """
     level = tf_logging.WARN if is_enabled else tf_logging.ERROR
     tf_logging.set_verbosity(level)
+
+
+def unflatten(vector, shapes):
+    i = 0
+    arrays = []
+    for shape in shapes:
+        size = np.prod(shape, dtype=np.int)
+        array = vector[i : (i + size)].reshape(shape)
+        arrays.append(array)
+        i += size
+    assert len(vector) == i, "Passed weight does not have the correct shape."
+    return arrays
+
+
+class TensorFlowVariables(object):
```

Comment:
Taken from https://github.com/ray-project/ray/blob/master/python/ray/experimental/tf_utils.py? Is there a better way to include this? Ray is probably pretty heavy weight.

Reply:
I included this from @harperj's https://github.com/Unity-Technologies/mlagents-distributed. Open to suggestions.

Reply:
They have an Apache 2.0 license https://github.com/ray-project/ray/blob/master/LICENSE; we can probably do this but will need to add attribution to our license too. You need to run this by Jeff and probably legal too.

Reply:
I'd recommend we not re-distribute this code from Ray.  It's something that is easy enough to re-implement.  Essentially we just want to capture all of `tf.trainable_variables()`, create an assignment op for each, and then capture them in a dictionary so we have the values to reassign later.

Here's a previous version from when I was working on parallel environments:
```python
class Policy:
    # ... other methods here ...
    def serialize_in_memory(self):
        with self.graph.as_default():
            trainable_vars = tf.trainable_variables()
            values = [v.eval(session=self.sess) for v in trainable_vars]
            return values

    def load_from_memory(self, values):
        with self.graph.as_default():
            assign_ops = []
            feed_dict = {}
            trainable_vars = tf.trainable_variables()
            weights = zip(trainable_vars, values)
            for var, value in weights:
                value = np.asarray(value)
                assign_placeholder = tf.placeholder(var.dtype, shape=value.shape)
                assign_op = tf.assign(var, assign_placeholder)
                assign_ops.append(assign_op)
                feed_dict[assign_placeholder] = value
            self.sess.run(assign_ops, feed_dict=feed_dict)
```

This implementation is missing some things it should have -- should re-use`tf.trainable_variables()` and not create new placeholder nodes each time since memory will grow unbounded.

Reply:
^- PS this is why I don't want anyone to make me delete my old branches :)

Reply:
Fixed this dependence.